programs/eden-tester/main.cpp

```diff
@@ -0,0 +1,1141 @@
+   }
+
+   template <typename T>
+   T unpack(const char* begin, size_t size)
+   {
+      fc::datastream<const char*> ds(begin, size);
+      T args;
+      fc::raw::unpack(ds, args);
+      return args;
+   }
+
+   template <typename T>
+   T unpack(span<const char> data)
+   {
+      return unpack<T>(data.begin(), data.size());
+   }
+
+   template <typename T>
+   T unpack(const eosio::input_stream& data)
+   {
+      return unpack<T>(data.pos, data.end - data.pos);
+   }
+
+   std::string span_str(span<const char> str) { return {str.data(), str.size()}; }
+
+   char* alloc(uint32_t cb_alloc_data, uint32_t cb_alloc, uint32_t size)
+   {
+      // todo: verify cb_alloc isn't in imports
+      if (state.backend.get_module().tables.size() < 0 ||
+          state.backend.get_module().tables[0].table.size() < cb_alloc)
+         throw std::runtime_error("cb_alloc is out of range");
+      auto result = state.backend.get_context().execute(  //
+          this, eosio::vm::jit_visitor(42), state.backend.get_module().tables[0].table[cb_alloc],
+          cb_alloc_data, size);
+      if (!result || !result->is_a<eosio::vm::i32_const_t>())
+         throw std::runtime_error("cb_alloc returned incorrect type");
+      char* begin = state.wa.get_base_ptr<char>() + result->to_ui32();
+      check_bounds(begin, size);
+      return begin;
+   }
+
+   template <typename T>
+   void set_data(uint32_t cb_alloc_data, uint32_t cb_alloc, const T& data)
+   {
+      memcpy(alloc(cb_alloc_data, cb_alloc, data.size()), data.data(), data.size());
+   }
+
+   void tester_abort() { throw std::runtime_error("called tester_abort"); }
+
+   void eosio_exit(int32_t) { throw std::runtime_error("called eosio_exit"); }
+
+   void eosio_assert_message(bool condition, span<const char> msg)
+   {
+      if (!condition)
+         throw ::assert_exception(span_str(msg));
+   }
+
+   void prints_l(span<const char> str) { std::cerr.write(str.data(), str.size()); }
+
+   void tester_get_arg_counts(wasm_ptr<uint32_t> argc, wasm_ptr<uint32_t> argv_buf_size)
+   {
+      uint32_t size = 0;
+      for (auto& a : state.args)
+         size += a.size() + 1;
+      *argc = state.args.size();
+      *argv_buf_size = size;
+   };
+
+   // uint8_t** argv, uint8_t* argv_buf
+   void tester_get_args(uint32_t argv, uint32_t argv_buf)
+   {
+      auto* memory = state.backend.get_context().linear_memory();
+      auto get_argv = [&]() -> uint32_t& { return *reinterpret_cast<uint32_t*>(memory + argv); };
+      auto get_argv_buf = [&]() -> uint8_t& {
+         return *reinterpret_cast<uint8_t*>(memory + argv_buf);
+      };
+
+      for (auto& a : state.args)
+      {
+         get_argv() = argv_buf;
+         argv += 4;
+         for (auto ch : a)
+         {
+            get_argv_buf() = ch;
+            ++argv_buf;
+         }
+         get_argv_buf() = 0;
+         ++argv_buf;
+      }
+   };
+
+   int32_t clock_gettime(int32_t id, void* data)
```

Comment:
This should probably be adjusted to implement the WASI interface.

Reply:
done